cip/2.testable/CIP2015-06-24-call-procedures.adoc

```diff
@@ -0,0 +1,236 @@
+An implementation may provide the user with a way to opt out of this behavior, however this must be done explicitly (e.g. via a configuration setting).
+
+==== Semantics of explicit argument passing
+
+In this scenario, arguments are provided explicitly as a sequence of expressions as required by the procedure's signature.
+It is an error if the number of provided arguments differs from the number of arguments required by the procedure signature.
+
+To call the procedure, all argument expressions are evaluated to argument values in order.
+It is an error if the argument values are incompatible with the argument types required by the procedure signature.
+
+==== Semantics of implicit argument passing
+
+In this scenario, arguments are provided implicitly via the parameters of the Cypher statement.
+
+To call the procedure, the argument values are obtained by using the parameter in scope with the same name as the procedure argument.
+If such a parameter does not exist, the argument value is taken to be `null`.
+It is an error, if the resulting argument values are incompatible with the argument types required by the procedure  signature.
+
+== What others do
+
+The https://docs.google.com/document/d/1t_MqE9e8cCcvW3YoHuP-aX2XSub15ksbeu2psRCTozY/edit?usp=sharing[stored procedures survey] is extremely comprehensive, examining how procedures are
+implemented and deployed as well as their API access mechanisms and usage.
+Products surveyed include PostgreSQL, MS SQL Server, Oracle, MySQL, MongoDB, Aerospike and Virtuoso.
+
+== Benefits to this proposal
+
+The benefits of having user-defined procedures is so that users would be able to implement algorithms and functionality
+which Cypher either cannot express or which cannot be executed efficiently by current Cypher implementations. Additionally, users may find procedures to be a useful mechanism to achieve good system design and code abstraction.
+
+== Caveats to this proposal
+
+Procedures are a powerful extension mechanism.
+Their introduction opens up new ways of using Cypher which over time may lead to suboptimal usage patterns and hard to read queries.
+The introduction of procedures therefore carries a risk of influencing the long term evolution of the language in a negative way.
+
+== Appendix: Calling modes
+
+.Calling modes
+[frame="topbot",cols="3l,2a,2a,1m", options="header,footer"]
+|===================================================================
+|Statement template        |Mode      |Argument Passing|New Bindings
+|.. CALL proc(..)          |In-Query  |Explicit        |None
+|.. CALL proc(..) YIELD .. |In-Query  |Explicit        |Some
+|CALL proc(..) YIELD ..    |Standalone|Explicit        |Some
+|CALL proc YIELD ..        |Standalone|Implicit        |Some
+|CALL proc(..)             |Standalone|Explicit        |All
+|CALL proc                 |Standalone|Implicit        |All
+|===================================================================
+
+Legend:
+* Mode
```

Comment:
This does not render as a bullet point list, I believe a blank line is needed before the first item in order to fix that.